### ISSUE TITLE
window.ipcRenderer.send is not a function

### ISSUE THREAD

**Issue by lostdesign** [nklayman/vue-cli-plugin-electron-builder#949]
**Describe the bug**
I went through https://nklayman.github.io/vue-cli-plugin-electron-builder/guide/security.html#node-integration and used a preload script to get the ipcRenderer on the window object. I am unable to send any events from vue to the background.js.

It will throw:
**To Reproduce**
`src/preload.js`
```js
import { ipcRenderer } from 'electron'
window.ipcRenderer = ipcRenderer
```

`background.js`
```js
ipcMain.on('select-dirs', async (event, arg) => {
  const result = await dialog.showOpenDialog(win, {
    properties: ['openDirectory']
  })
  console.log('directories selected', result.filePaths)
  win.webContents.send('select-dirs-reply', result.filePaths)
})
```

`settings.vue`
```html
<template>
  <main>
    <h1>Settings</h1>
    <input type="file" @click.prevent="openDialog"/>
  </main>
</template>

<script>
export default {
  methods: {
    openDialog() {
      window.ipcRenderer.send('select-dir')
    }
  },
}
</script>
```

**Expected behavior**
The IPC event should fire.

**Environment (please complete the following information):**
  System:
    OS: macOS 10.15.5
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
  Binaries:
    Node: 14.2.0 - /usr/local/bin/node
    Yarn: Not Found
    npm: 6.14.4 - /usr/local/bin/npm
  Browsers:
    Chrome: 84.0.4147.105
    Edge: Not Found
    Firefox: Not Found
    Safari: 13.1.1
  npmPackages:
    @vue/babel-helper-vue-jsx-merge-props:  1.0.0 
    @vue/babel-plugin-transform-vue-jsx:  1.1.2 
    @vue/babel-preset-app:  4.4.6 
    @vue/babel-preset-jsx:  1.1.2 
    @vue/babel-sugar-functional-vue:  1.1.2 
    @vue/babel-sugar-inject-h:  1.1.2 
    @vue/babel-sugar-v-model:  1.1.2 
    @vue/babel-sugar-v-on:  1.1.2 
    @vue/cli-overlay:  4.4.6 
    @vue/cli-plugin-babel: ~4.4.0 => 4.4.6 
    @vue/cli-plugin-eslint: ~4.4.0 => 4.4.6 
    @vue/cli-plugin-router: ~4.4.0 => 4.4.6 
    @vue/cli-plugin-vuex:  4.4.6 
    @vue/cli-service: ~4.4.0 => 4.4.6 
    @vue/cli-shared-utils:  4.4.6 
    @vue/component-compiler-utils:  3.2.0 
    @vue/preload-webpack-plugin:  1.1.2 
    @vue/web-component-wrapper:  1.2.0 
    eslint-plugin-vue: ^6.2.2 => 6.2.2 
    vue: ^2.6.11 => 2.6.11 
    vue-cli-plugin-electron-builder: ~2.0.0-rc.4 => 2.0.0-rc.4 
    vue-eslint-parser:  7.1.0 
    vue-hot-reload-api:  2.3.4 
    vue-loader:  15.9.3 
    vue-router: ^3.2.0 => 3.4.0 
    vue-style-loader:  4.1.2 
    vue-template-compiler: ^2.6.11 => 2.6.11 
    vue-template-es2015-compiler:  1.9.1 
  npmGlobalPackages:
    @vue/cli: 4.4.6

*Stacktrace*
```
webpack-internal:///./node_modules/webpack/hot/log.js:24 [HMR] Waiting for update signal from WDS...
vue.runtime.esm.js?2b0e:619 [Vue warn]: Error in v-on handler: "TypeError: Cannot read property 'send' of undefined"

found in

---> <Settings> at src/views/Settings.vue
       <App> at src/App.vue
         <Root>
warn @ vue.runtime.esm.js?2b0e:619
logError @ vue.runtime.esm.js?2b0e:1884
globalHandleError @ vue.runtime.esm.js?2b0e:1879
handleError @ vue.runtime.esm.js?2b0e:1839
invokeWithErrorHandling @ vue.runtime.esm.js?2b0e:1862
invoker @ vue.runtime.esm.js?2b0e:2179
original._wrapper @ vue.runtime.esm.js?2b0e:6917
vue.runtime.esm.js?2b0e:1888 TypeError: Cannot read property 'send' of undefined
    at VueComponent.openDialog (Settings.vue?e12e:12)
    at invokeWithErrorHandling (vue.runtime.esm.js?2b0e:1854)
    at HTMLInputElement.invoker (vue.runtime.esm.js?2b0e:2179)
    at HTMLInputElement.original._wrapper (vue.runtime.esm.js?2b0e:6917)
logError @ vue.runtime.esm.js?2b0e:1888
globalHandleError @ vue.runtime.esm.js?2b0e:1879
handleError @ vue.runtime.esm.js?2b0e:1839
invokeWithErrorHandling @ vue.runtime.esm.js?2b0e:1862
invoker @ vue.runtime.esm.js?2b0e:2179
original._wrapper @ vue.runtime.esm.js?2b0e:6917
vue.runtime.esm.js?2b0e:619 [Vue warn]: Error in v-on handler: "TypeError: Cannot read property 'send' of undefined"

found in

---> <Settings> at src/views/Settings.vue
       <App> at src/App.vue
         <Root>
warn @ vue.runtime.esm.js?2b0e:619
logError @ vue.runtime.esm.js?2b0e:1884
globalHandleError @ vue.runtime.esm.js?2b0e:1879
handleError @ vue.runtime.esm.js?2b0e:1839
invokeWithErrorHandling @ vue.runtime.esm.js?2b0e:1862
invoker @ vue.runtime.esm.js?2b0e:2179
original._wrapper @ vue.runtime.esm.js?2b0e:6917
vue.runtime.esm.js?2b0e:1888 TypeError: Cannot read property 'send' of undefined
    at VueComponent.openDialog (Settings.vue?e12e:12)
    at invokeWithErrorHandling (vue.runtime.esm.js?2b0e:1854)
    at HTMLInputElement.invoker (vue.runtime.esm.js?2b0e:2179)
    at HTMLInputElement.original._wrapper (vue.runtime.esm.js?2b0e:6917)
logError @ vue.runtime.esm.js?2b0e:1888
globalHandleError @ vue.runtime.esm.js?2b0e:1879
handleError @ vue.runtime.esm.js?2b0e:1839
invokeWithErrorHandling @ vue.runtime.esm.js?2b0e:1862
invoker @ vue.runtime.esm.js?2b0e:2179
original._wrapper @ vue.runtime.esm.js?2b0e:6917
```

**Comment by nklayman** [comment]
Can you post the config passed to `new BrowserWindow` in your `background.js`?

**Comment by lostdesign** [comment]
Unfortunately I deleted the project and went with a custom Webpack config instead. The nodeIntegration was set to true, if that is what you wanted to know.

**Comment by nklayman** [comment]
I wanted to see what path you used for the preload file. Do you remember what you used?

**Comment by ihewro** [comment]
```
function createWindow(devPath, prodPath,w=800,h=600, haveFrame = true) {
  // Create the browser window.
  const path = require("path");

  let window = new BrowserWindow({
    width: w,
    height: h ,
    frame:haveFrame,
    preload: path.join(__dirname, 'preload.js'),
    contextIsolation:false
  })

```

I have the same problem

**Comment by lostdesign** [comment]
@ihewro i've fixed that for me by not using preload at all. I simply import the `ipcRenderer` in the frontend via `nodeIntegration: true` see https://github.com/muc-dev/linked/blob/9becd92f2f167928f578741896c60ad4b04ec5a1/vue.config.js#L19 and https://github.com/muc-dev/linked/blob/master/src/mixins/file.js

I cannot quite remember what it was or why I got it to work the way I am using it now, but it made sense and is fine for my use-case. 


@nklayman regarding the path, it's pretty much what @ihewro posted above I'd say.

**Comment by nklayman** [comment]
@ihewro you may need to disable [contextIsolation](https://www.electronjs.org/docs/tutorial/context-isolation). See [this page of the docs](https://nklayman.github.io/vue-cli-plugin-electron-builder/guide/security.html#node-integration) for more details. I'm going to close this issue as the original problem seems to have been resolved, but you can open a new issue if the guide I linked doesn't work.

**Comment by abulka** [comment]
@nklayman I am using electron `13.6.3` and found that I had to set `contextIsolation` to `false` inside background.js createWindow() webPreferences otherwise assigning `window.ipcRenderer = ipcRenderer` will work _only_ inside the `preload.js` module, and annoyingly, `window.ipcRenderer` reverts to being `undefined` again elsewhere - where you'd want to use it to make interprocess calls.  P.S. My project has `nodeIntegration` as the recommended default of `false`.

Possibly the documentation page you refer to above, and most certainly the [preload documentation](https://nklayman.github.io/vue-cli-plugin-electron-builder/guide/guide.html#preload-files) and even possibly the Common Issues section should probably mention this, as most developers building vue-electron apps will want to use interprocess communication and will need to use the `preload.js` technique - which, with the current defaults will fail because the default `contextIsolation: !process.env.ELECTRON_NODE_INTEGRATION` evaluates to `true`, thus mysteriously (at least for me till I figured it out) breaking the `preload.js` technique.
